File: src/main/java/no/nav/syfo/services/FastlegeService.java
package no.nav.syfo.services;

import lombok.extern.slf4j.Slf4j;
import no.nav.syfo.consumer.fastlege.FastlegeConsumer;
import no.nav.syfo.consumer.fastlege.PraksisInfo;
import no.nav.syfo.consumer.pdl.*;
import no.nav.syfo.domain.*;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import javax.inject.Inject;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;

import static java.time.LocalDate.now;
import static java.util.Optional.ofNullable;
import static java.util.stream.Collectors.toList;
import static no.nav.syfo.util.StringUtil.lowerCapitalize;

@Slf4j
@Service
public class FastlegeService {
    private final PdlConsumer pdlConsumer;
    private final FastlegeConsumer fastlegeConsumer;

    @Inject
    public FastlegeService(
            final PdlConsumer pdlConsumer,
            final FastlegeConsumer fastlegeConsumer) {
        this.pdlConsumer = pdlConsumer;
        this.fastlegeConsumer = fastlegeConsumer;
    }

    @Cacheable(value = "fastlege")
    public Optional<Fastlege> hentBrukersFastlege(String brukersFnr) {
        return hentBrukersFastleger(brukersFnr).stream()
                .filter(isAktiv())
                .findFirst();
    }

    @Cacheable(value = "fastlege")
    public List<Fastlege> hentBrukersFastleger(String brukersFnr) {
        try {
            Optional<PdlHentPerson> maybePerson = Optional.ofNullable(pdlConsumer.person(brukersFnr));
            Pasient pasient = toPasient(maybePerson);
            List<Fastlege> fastlegeListe = fastlegeConsumer.getFastleger(brukersFnr);
            return fastlegeListe.stream()
                    .map(fastlege -> fastlege
                            .pasient(new Pasient()
                                    .fnr(brukersFnr)
                                    .fornavn(pasient.fornavn())
                                    .mellomnavn(pasient.mellomnavn())
                                    .etternavn(pasient.etternavn()))
                            .foreldreEnhetHerId(hentForeldreEnhetHerId(fastlege).orElse(null))
                    ).collect(toList());
        } catch (RuntimeException e) {
            log.error("Søkte opp og fikk en feil fra fastlegetjenesten fordi tjenesten er nede", e);
            throw e;
        }
    }

    private Pasient toPasient(Optional<PdlHentPerson> maybePerson) {
        Pasient pasient = new Pasient();
        if (maybePerson.isPresent()) {
            Optional<PdlPerson> pdlHentPerson = Optional.ofNullable(maybePerson.get().getHentPerson());
            if (pdlHentPerson.isPresent()) {
                List<PdlPersonNavn> nameList = pdlHentPerson.get().getNavn();
                if (nameList.isEmpty()) {
                    return pasient;
                } else {
                    PdlPersonNavn personNavn = nameList.get(0);
                    pasient.fornavn = lowerCapitalize(personNavn.getFornavn());
                    pasient.mellomnavn = lowerCapitalize(personNavn.getMellomnavn());
                    pasient.etternavn = lowerCapitalize(personNavn.getEtternavn());
                }
            }
        }
        return pasient;
    }

    private Optional<Integer> hentForeldreEnhetHerId(Fastlege fastlege) {
        return ofNullable(fastlege.herId())
                .map(fastlegeConsumer::getPraksisInfo)
                .map(PraksisInfo::getForeldreEnhetHerId);
    }

    private static Predicate<Fastlege> isAktiv() {
        return fastlege -> fastlege.pasientforhold().fom().isBefore(now()) && fastlege.pasientforhold().tom().isAfter(now());
    }
}

File: src/main/kotlin/no/nav/syfo/fastlege/FastlegeService.kt
package no.nav.syfo.fastlege

import no.nav.syfo.application.cache.RedisStore
import no.nav.syfo.client.pdl.PdlClient
import no.nav.syfo.client.pdl.PdlHentPerson
import no.nav.syfo.fastlege.domain.*
import no.nav.syfo.fastlege.ws.adresseregister.AdresseregisterClient
import no.nav.syfo.fastlege.ws.fastlegeregister.FastlegeInformasjonClient
import no.nav.syfo.util.*
import org.slf4j.LoggerFactory
import java.time.LocalDate

class FastlegeService(
    private val pdlClient: PdlClient,
    private val fastlegeClient: FastlegeInformasjonClient,
    private val adresseregisterClient: AdresseregisterClient,
    private val cache: RedisStore,
) {
    suspend fun hentBrukersFastleger(
        personIdent: PersonIdent,
    ): List<Fastlege> =
        try {
            val cacheKey = "fastleger-$personIdent"
            val cachedValue = cache.getListObject<Fastlege>(cacheKey)
            if (cachedValue != null) {
                COUNT_CALL_FASTLEGER_CACHE_HIT.increment()
                cachedValue
            } else {
                val maybePerson = pdlClient.person(personIdent)
                val pasient = toPasient(personIdent, maybePerson)
                fastlegeClient.hentBrukersFastleger(personIdent).map { fastlege ->
                    fastlege.copy(
                        pasient = Pasient(
                            fnr = personIdent.value,
                            fornavn = pasient?.fornavn ?: "",
                            mellomnavn = pasient?.mellomnavn,
                            etternavn = pasient?.etternavn ?: "",
                        ),
                        foreldreEnhetHerId = hentForeldreEnhetHerId(fastlege.herId),
                    )
                }.also {
                    COUNT_CALL_FASTLEGER_CACHE_MISS.increment()
                    cache.setObject(cacheKey, it, 24 * 3600)
                }
            }
        } catch (e: RuntimeException) {
            log.error("Søkte opp og fikk en feil fra fastlegetjenesten fordi tjenesten er nede", e)
            throw e
        }

    suspend fun hentBrukersFastlege(
        personIdent: PersonIdent,
    ): Fastlege? {
        return hentBrukersFastleger(
            personIdent = personIdent,
        ).aktiv()
    }

    suspend fun hentBrukersFastlegevikar(
        personIdent: PersonIdent,
    ): Fastlege? {
        return hentBrukersFastleger(
            personIdent = personIdent,
        ).vikar()
    }

    fun hentBehandlereForKontor(
        kontorHerId: Int,
    ): BehandlerKontor? =
        adresseregisterClient.hentBehandlereForKontor(kontorHerId)

    private fun toPasient(
        personIdent: PersonIdent,
        maybePerson: PdlHentPerson?,
    ): Pasient? {
        return maybePerson?.hentPerson?.let { pdlPerson ->
            pdlPerson.navn.firstOrNull()?.let { pdlPersonNavn ->
                Pasient(
                    fnr = personIdent.value,
                    fornavn = pdlPersonNavn.fornavn.lowerCapitalize(),
                    mellomnavn = pdlPersonNavn.mellomnavn?.lowerCapitalize(),
                    etternavn = pdlPersonNavn.etternavn.lowerCapitalize(),
                )
            }
        }
    }

    private fun hentForeldreEnhetHerId(
        fastlegeHerId: Int?,
    ): Int? {
        return fastlegeHerId?.let { herId ->
            adresseregisterClient.hentPraksisInfoForFastlege(
                herId = herId,
            )?.foreldreEnhetHerId
        }
    }

    private fun List<Fastlege>.aktiv(): Fastlege? {
        return this.filter { fastlege ->
            fastlege.relasjon.kodeVerdi == RelasjonKodeVerdi.FASTLEGE.kodeVerdi
        }.maxByOrNull { fastlege ->
            fastlege.gyldighet.fom
        }
    }

    private fun List<Fastlege>.vikar(): Fastlege? {
        val today = LocalDate.now()
        return this.filter { fastlege ->
            fastlege.relasjon.kodeVerdi == RelasjonKodeVerdi.VIKAR.kodeVerdi
        }.filter { vikar ->
            vikar.gyldighet.fom.isBeforeOrEqual(today) && vikar.gyldighet.tom.isAfterOrEqual(today)
        }.maxByOrNull { vikar ->
            vikar.gyldighet.fom
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(FastlegeService::class.java)
    }
}
